### DISCUSSION TITLE
pdns-recursor is returning rcode="3"

### DISCUSSION THREAD

**Discussion by phancong0897** [PowerDNS/pdns#14627] [Q&A]
I'm using the dns model dnsdist --> pdns-recursor --> pdns , I'm having a situation where the first domain lookup returns nxdomain error results, but when accessing the second time, the domain resolves normally. , 2 consecutive queries, no time delay, I see pdns-recursor is returning rcode="3" and there is no query log from pdns-recurosr to pdns, has anyone seen it or just me, here is my configuration

**Comment by phancong0897** [comment]
pdns-distributes-queries=yes
reuseport=no
max-packetcache-entries=4000000
packetcache-negative-ttl=1
packetcache-shards=2048
packetcache-servfail-ttl=60
max-cache-bogus-ttl=60
max-negative-ttl=60
max-udp-queries-per-round=50000
record-cache-locked-ttl-perc=10
packetcache-ttl=86400
minimum-ttl-override=86400
refresh-on-ttl-perc=10
event-trace-enabled=2
unique-response-log=yes
serve-stale-extensions=120
network-timeout=3000
stack-cache-size=512
daemon=no
threads=8
distributor-threads=8
allow-from=0.0.0.0/0, ::/0
log-common-errors=yes
non-local-bind=yes
local-address=0.0.0.0
# dnssec
dnssec=off
disable-syslog=no
quiet=no

**Comment by omoerbeek** [comment]
1. Can you start the recursor with --trace, do your initial and second query and then share the log? That probably will shed a light on what is happening.

2. On a side note: I have some thoughts about your setting: e.g. `distributor-threads=8` makes no sense if you have `threads=8` and `disable-syslog` and `quiet` have nothing to do with dnssec. running with `quiet=no` can help to study what is going on, but you do not want that in a production environment. I'd suggest you start running with default settings and only change settings if you have a clear indication it's needed or will improve things in your setup.

**Comment by phancong0897** [comment]
i used to be silent=yes, however i want to achieve optimal performance. nxdomain error is not continuous, but in 1 hour there are 5-7 random domains that are faulty, it is not a continuous domain. pdsn-recursor log is the second return of rcode=3, the second return of rcode=0. currently my dns is under load of about 3000 requests per second, with 8 CPUs 8 GBRAM, is my above configuration optimized, can you optimize for me to get the best performance of dnsdist and pdns-recursor.
Apart from the above error, i have not encountered any other errors, i have a headache because of it.
